### PR TITLE
Dict: Support for objects as values; Adds ToJSON()

### PR DESCRIPTION
**Pull request by nseam**
Refs #88, #89

**All work is done**

### CODE REVIEW

File: tests/DictTest.mq5 [hunk 1]
```diff
@@ -58,5 +58,29 @@ int OnInit() {
...
+  dict3.Set(1, &dict2);
+  Dict<int, string>* dict2_ref = dict3.GetByKey(1);
+  assertTrueOrFail(dict2_ref != NULL, "Dict should return non-NULL pointer to the dict2 object!");
+  assertTrueOrFail(dict2_ref.GetByKey(1) == "a", "Incorrect value read from dict2 object. Expected 'a' for key 1!");
+  dict2.Set(1, "d");
+  assertTrueOrFail(dict2_ref.GetByKey(1) == "d", "Reference to dict2 doesn't point to the dict2 object, but rather to a copy of dict2. It is wrong!");
+  dict2_ref.Unset(1);
+  assertTrueOrFail(dict2_ref.KeyExists(1) == false, "Dict should'nt contain key 1 as it was unset!");
+
+  // Example 3. Dictionary of other dictionaries.
```
Comment (kenorb) [inline tests/DictTest.mq5]: It's 4th example.

File: DictObject.mqh [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    // We can't shrink to less than number of already used slots.
+      return;
+    }
+
+    SlotsRef new_slots;
+
+    ArrayResize(new_slots.slots, new_size);
+
+    // Copies entire array of slots into new array of slots. Hashes will be rehashed.
+    for (unsigned int i = 0; i < (unsigned int) ArraySize(_slots_ref.slots); ++i) {
```
Comment (kenorb) [inline DictObject.mqh]: Move `unsigned int i` declaration to separate like, as g++ doesn't like it.

File: JSON.mqh [hunk 3]
```diff
@@ -0,0 +1,28 @@
+#define JSON_INDENTATION 2
+
+#include "Dict.mqh"
```
Comment (kenorb) [inline JSON.mqh]: Dict not used in this class.

File: DictSimpleValues.mqh [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    // We can't shrink to less than number of already used slots.
+      return;
+    }
+
+    SlotsRef new_slots;
+
+    ArrayResize(new_slots.slots, new_size);
+
+    // Copies entire array of slots into new array of slots. Hashes will be rehashed.
+    for (unsigned int i = 0; i < (unsigned int) ArraySize(_slots_ref.slots); ++i) {
```
Comment (kenorb) [inline DictSimpleValues.mqh]: Move declaration of `i` to separate line to C++ compatibility.

File: DictBase.mqh [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+
+public:
+
+  DictBase()
+  {
+    _current_id = 0;
+    _num_used   = 0;
+  }
+  
+  uint GetSlotCount() {
```
Comment (kenorb) [inline DictBase.mqh]: Replace all `uint` to `unsigned int`.

File: DictSimpleValues.mqh [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+  {
+    string _rows[], _row[];
+    int _rows_count = StringSplit(_data, StringGetCharacter(_dlm, 0), _rows);
+    int _i;
+    if (_rows_count > 0) {
+      Resize(_rows_count);
+      for (_i = 0; _i < _rows_count; _i++) {
+        int _row_count = StringSplit(_rows[_i], StringGetCharacter("=", 0), _row);
+        if (_row_count >= 2) {
+          //Set((K) _row[0], (V) _row[1]);
```
Comment (kenorb) [inline DictSimpleValues.mqh]: //todo ?

File: DictBase.mqh [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+    return value.ToJSON(indent);
+  }
+  
+  virtual string ToJSON(uint indentation = 2) {
+    string json = "{\n";
+    
+    uint numSlots = GetSlotCount();
+    bool alreadyStarted = false;
+    
+    for (uint i = 0; i < numSlots; ++i)
```
Comment (kenorb) [inline DictBase.mqh]: Move declaration of `i` to separate line.

File: DictBase.mqh [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+
+  // Number of used slots.
+  unsigned int _num_used;
+
+public:
+
+  /**
+  * Represents a single item in the hash table.
+  */
+  struct Slot
```
Comment (kenorb) [inline DictBase.mqh]: This struct should be moved before the class definition.

File: DictBase.mqh [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+  {
+    bool has_key;  // Whether slot has key (false if slot is hashless).
+    bool is_used;  // Whether slot is currently in use.
+    bool was_used; // Whether slots was in use and now is in use or was emptied.
+    K    key;      // Key used to store value.
+    V    value;    // Value stored.
+  
+    Slot ()
+    {
+      is_used = was_used = has_key = false;
```
Comment (kenorb) [inline DictBase.mqh]: Use initialization list, e.g. `Slot : is_used(false), was_used (false), ...`.

File: DictBase.mqh [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+  struct SlotsRef
+  {
+     Slot slots[];
+  };
+
+public:
+
+  DictBase()
+  {
+    _current_id = 0;
```
Comment (kenorb) [inline DictBase.mqh]: Use initialization list.

File: DictObject.mqh (hunk 2 above)
Comment (nseam) [reply DictObject.mqh]: Could you make sure it really happens in G++?

**Pull request merged**